### PR TITLE
[1770] LA funded places

### PR DESCRIPTION
**Pull request by tariqhamid**
### Context
Provide independent schools (ISS) context for LA funded places

### Changes proposed in this pull request
Addition of LA Funded Type and interstitial LA Funded page when user logs in

### Guidance to review
The login user journey has been changed in 
  app/controllers/school/welcome_wizard_controller.rb 
  app/models/user.rb
  app/services/interstitial_picker.rb

The LA Funded place follows the journey given in the prototype: https://ghwt-prototype.herokuapp.com/school/welcome

### CODE REVIEW

**Comment by asmega** [comment]
Can we squash these commits, the only merge commits in the repository should be those of pull requests

File: app/models/local_authority.rb [hunk 1]
```diff
@@ -15,4 +15,31 @@ class LocalAuthority < ResponsibleBody
   }
 
   validates :organisation_type, presence: true
+
+  def la_funded_place
+    schools.la_funded_place_establishment_type&.first
```
Comment (nigel-lowry) [inline app/models/local_authority.rb:20]: If there is nothing in the array and this returns a `nil` does this cause you a problem? Is there a spec for that case?

File: app/models/local_authority.rb [hunk 2]
```diff
@@ -15,4 +15,31 @@ class LocalAuthority < ResponsibleBody
...
+  end
+
+  def create_la_funded_places!(urn:, device_allocation: 0, router_allocation: 0, extra_args: {})
+    return unless la_funded_place.nil?
```
Comment (nigel-lowry) [inline app/models/local_authority.rb:24]: I'd probably make that `return if la_funded_place.present?`

File: spec/factories/schools.rb [hunk 3]
```diff
@@ -6,12 +6,17 @@
...
+      establishment_type { 'la_funded_place' }
+    end
+
     association :responsible_body, factory: %i[local_authority trust].sample
     urn { Faker::Number.unique.number(digits: 6) }
     sequence(:name) { |n| "#{Faker::Educator.secondary_school}-#{n}" }
     computacenter_reference { Faker::Number.number(digits: 8) }
     phase { School.phases.values.sample }
-    establishment_type { School.establishment_types.values.sample }
+    establishment_type { (School.establishment_types.values - %w[la_funded_place]).sample }
```
Comment (nigel-lowry) [inline spec/factories/schools.rb:19]: Is it worth using `%w` when there's only one element in the array? I'd probably have just put that one item in quotes.

File: spec/factories/users.rb [hunk 4]
```diff
@@ -43,7 +43,7 @@
 
     factory :single_academy_trust_user do
       association :responsible_body, factory: %i[trust single_academy_trust]
-      school { build(:school, :academy, responsible_body: responsible_body) }
+      schools { build_list(:school, 1, :academy, responsible_body: responsible_body) }
```
Comment (nigel-lowry) [inline spec/factories/users.rb:46]: Is it worth using `build_list` to build a list of one element?

**Comment by asmega** [comment]
Can we delete the following line https://github.com/DFE-Digital/get-help-with-tech/pull/1550/files#diff-97e661aa6e51c41e8d02b50ee1749b96cd4042ad1eb8a88fca9c9f68e2ce1f42R74 since the setup is now slightly different

**Comment by tariqhamid** [comment]
> Can we delete the following line https://github.com/DFE-Digital/get-help-with-tech/pull/1550/files#diff-97e661aa6e51c41e8d02b50ee1749b96cd4042ad1eb8a88fca9c9f68e2ce1f42R74 since the setup is now slightly different

Which line?

**Comment by asmega** [comment]
> > Can we delete the following line https://github.com/DFE-Digital/get-help-with-tech/pull/1550/files#diff-97e661aa6e51c41e8d02b50ee1749b96cd4042ad1eb8a88fca9c9f68e2ce1f42R74 since the setup is now slightly different
> 
> Which line?

Sorry, I updated my comment with correct reference

**Comment by tariqhamid** [comment]
> > > Can we delete the following line https://github.com/DFE-Digital/get-help-with-tech/pull/1550/files#diff-97e661aa6e51c41e8d02b50ee1749b96cd4042ad1eb8a88fca9c9f68e2ce1f42R74 since the setup is now slightly different
> > 
> > 
> > Which line?
> 
> Sorry, I updated my comment with correct reference

Deleting that line causes multiple rspec errors:

  1) SchoolWelcomeWizard#update_step! when the step is allocation moves to the devices_you_can_order step
     Failure/Error: wizard.allocation!
     
     NoMethodError:
       undefined method `allocation!' for nil:NilClass
     # ./spec/models/school_welcome_wizard_spec.rb:16:in `block (4 levels) in <top (required)>'
     # /Users/tariq/.rvm/gems/ruby-2.7.2/gems/webmock-3.12.2/lib/webmock/rspec.rb:37:in `block (2 levels) in <top (required)>'

File: spec/factories/schools.rb (hunk 3 above)
Comment (tariqhamid) [reply spec/factories/schools.rb:19]: rubocop added the %w

File: spec/factories/users.rb (hunk 4 above)
Comment (tariqhamid) [reply spec/factories/users.rb:46]: yes, because using build alone the rspec fails

File: app/models/local_authority.rb (hunk 1 above)
Comment (tariqhamid) [reply app/models/local_authority.rb:20]: tonyheadford 1 hour ago Contributor
The & is superfluous here

Suggested change 
    schools.la_funded_place_establishment_type&.first
    schools.la_funded_place_establishment_type.first

[the remark this answers is not in the document]
Comment (tariqhamid) [reply app/models/local_authority.rb:20]: The & was suggest by @nigel-lowry 

removed &

File: app/models/local_authority.rb (hunk 2 above)
Comment (tariqhamid) [reply app/models/local_authority.rb]: fixed

File: app/models/local_authority.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (tariqhamid) [reply app/models/local_authority.rb]: fixed

File: spec/features/school/navigate_school_welcome_wizard_spec.rb [hunk 5]
```diff
@@ -172,6 +187,18 @@ def when_i_sign_in_for_the_first_time
     visit validate_token_url_for(@user)
   end
 
+  def then_i_state_funded_interstitial
```
[the remark this answers is not in the document]
Comment (tariqhamid) [reply spec/features/school/navigate_school_welcome_wizard_spec.rb]: fixed

**Comment by tariqhamid** [comment]
> Can we delete the following line https://github.com/DFE-Digital/get-help-with-tech/blob/main/spec/factories/users.rb#L74 since the setup is now slightly different

fixed

File: app/models/local_authority.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (tariqhamid) [reply app/models/local_authority.rb]: fixed

**Review by nigel-lowry — APPROVED** [review]

**Review by tonyheadford — APPROVED** [review]

**Pull request merged**